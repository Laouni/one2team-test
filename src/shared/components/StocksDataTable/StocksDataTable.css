.table {
  min-width: 5%;
}

.content-editable {
  padding: 1rem;
}

.content-editable:hover {
  background: #f7f7f7;
}

.content-editable:focus {
  background: #fcf8e1;
  outline: 0;
}

.paper {
  width: 100%;
  overflow-x: auto;
}

.root {
  width: 60%;
}
