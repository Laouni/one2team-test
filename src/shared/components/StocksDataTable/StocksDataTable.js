// @flow

import React from 'react';
import { connect } from 'react-redux';
import ContentEditable from 'react-contenteditable';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import { setStocks } from '../../redux/actions/StocksActions';

import './StocksDataTable.css';

type PropTypes = {
  data: Array<Object>,
  updateStocks: Function,
};

type StateTypes = {
  stocks: Array<Object>,
  row: Object,
};

/**
 * StocksDataTable
 */
class StocksDataTable extends React.Component<PropTypes, StateTypes> {
  /**
   * constructor
   * @param props
   */
  constructor(props: PropTypes) {
    super(props);
    this.setState({
      stocks: props.data,
      row: { CAC40: '', NASDAQ: '' },
    });
  }

  /**
   * highlight select cell
   */
  highlightAll = () => {
    setTimeout(() => {
      document.execCommand('selectAll', false, null);
    }, 0);
  };

  /**
   * check if it's a number
   * @param event
   */
  validateNumber = event => {
    const keyCode = event.keyCode || event.which;
    console.log(`My keCode ${keyCode}`);
    const string = String.fromCharCode(keyCode);
    console.log(`My string ${string}`);
    const regex = /[0-9,]|\./;

    if (!regex.test(string)) {
      event.returnValue = false;
      if (event.preventDefault) event.preventDefault();
    }
  };

  /**
   * paste as plain text
   * @param event
   */
  pasteAsPlainText = event => {
    event.preventDefault();

    const text = event.clipboardData.getData('text/plain');
    document.execCommand('insertHTML', false, text);
  };

  /**
   * handle content editable update
   * @param event
   */
  handleContentEditableUpdate = event => {
    const { data, updateStocks } = this.props;

    const {
      currentTarget: {
        dataset: { row, column },
      },
      target: { value },
    } = event;

    // eslint-disable-next-line radix
    const updatedRow = data.filter((item, i) => parseInt(i) === parseInt(row))[0];
    updatedRow[column] = value;
    const stocks = data.map((item, i) => (item[column] === row ? updatedRow : item));

    this.setState({ stocks });
    updateStocks(stocks);
  };

  /**
   * render
   * @returns {HTMLElement}
   */
  render() {
    const { data } = this.props;

    return (
      <div className="root">
        <Paper className="paper">
          <Table className="table" size="small" aria-label="a dense table">
            <TableHead>
              <TableRow>
                <TableCell>Number</TableCell>
                <TableCell>CAC40</TableCell>
                <TableCell>NASDAQ</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              {data.map((row, i) => (
                <TableRow key={row.index}>
                  <TableCell>{row.index}</TableCell>
                  <TableCell>
                    <ContentEditable
                      html={row.CAC40}
                      data-column="CAC40"
                      data-row={i}
                      className="content-editable"
                      onKeyPress={this.validateNumber}
                      onPaste={this.pasteAsPlainText}
                      onFocus={this.highlightAll}
                      onChange={this.handleContentEditableUpdate}
                    />
                  </TableCell>
                  <TableCell>
                    <ContentEditable
                      html={row.NASDAQ}
                      data-column="NASDAQ"
                      data-row={i}
                      className="content-editable"
                      onKeyPress={this.validateNumber}
                      onPaste={this.pasteAsPlainText}
                      onFocus={this.highlightAll}
                      onChange={this.handleContentEditableUpdate}
                    />
                  </TableCell>
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </Paper>
      </div>
    );
  }
}

/**
 * @private
 * @param {Object} state
 * @returns {{}}
 */
const mapStateToProps = (state: Object) => ({
  ...state,
});

/**
 * @private
 * @param {Object} dispatch
 * @returns {{actions: *}}
 */
const mapDispatchToProps = {
  updateStocks: setStocks,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(StocksDataTable);
