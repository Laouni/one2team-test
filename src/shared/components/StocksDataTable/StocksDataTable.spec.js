import React from 'react';
import renderer from 'react-test-renderer';
import StocksDataTable from './StocksDataTable';

const defaultProps = {
  data: [
    {
      index: 1,
      NASDAQ: 5.67,
      CAC40: 12.45,
    },
    {
      index: 2,
      NASDAQ: 15.57,
      CAC40: 10.43,
    },
    {
      index: 3,
      NASDAQ: 18.61,
      CAC40: 22.25,
    },
    {
      index: 4,
      NASDAQ: 9.87,
      CAC40: 11.48,
    },
    {
      index: 5,
      NASDAQ: 35.67,
      CAC40: 22.95,
    },
  ],
};

describe('StocksDataTable compoent', () => {
  it('should render correctly without props', () => {
    const component = renderer.create(<StocksDataTable />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with props', () => {
    const component = renderer.create(<StocksDataTable data={defaultProps.data} />);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
