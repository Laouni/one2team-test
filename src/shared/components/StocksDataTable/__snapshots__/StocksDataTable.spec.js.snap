// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StocksDataTable compoent should render correctly with props 1`] = `
<div
  className="root"
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 paper MuiPaper-rounded"
  >
    <table
      aria-label="a dense table"
      className="MuiTable-root table"
    >
      <thead
        className="MuiTableHead-root"
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
            scope="col"
          >
            Number
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
            scope="col"
          >
            CAC40
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-head"
            scope="col"
          >
            NASDAQ
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
      >
        <tr
          className="MuiTableRow-root"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            1
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 12.45,
                }
              }
              data-column="CAC40"
              data-row={0}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 5.67,
                }
              }
              data-column="NASDAQ"
              data-row={0}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            2
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 10.43,
                }
              }
              data-column="CAC40"
              data-row={1}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 15.57,
                }
              }
              data-column="NASDAQ"
              data-row={1}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            3
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 22.25,
                }
              }
              data-column="CAC40"
              data-row={2}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 18.61,
                }
              }
              data-column="NASDAQ"
              data-row={2}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            4
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 11.48,
                }
              }
              data-column="CAC40"
              data-row={3}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 9.87,
                }
              }
              data-column="NASDAQ"
              data-row={3}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
        </tr>
        <tr
          className="MuiTableRow-root"
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            5
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 22.95,
                }
              }
              data-column="CAC40"
              data-row={4}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-sizeSmall MuiTableCell-body"
          >
            <div
              className="content-editable"
              contentEditable={true}
              dangerouslySetInnerHTML={
                Object {
                  "__html": 35.67,
                }
              }
              data-column="NASDAQ"
              data-row={4}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onPaste={[Function]}
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
